src/type/validate.js

```diff
@@ -643,3 +696,126 @@ function getDeprecatedDirectiveNode(
     (node) => node.name.value === GraphQLDeprecatedDirective.name,
   );
 }
+
+/**
+ * Coerce an internal JavaScript value given a GraphQL Input Type.
+ */
+function validateDefaultValue(
+  inputValue: mixed,
+  type: GraphQLInputType,
+  path?: Path,
+): Array<string> {
+  if (isNonNullType(type)) {
+    if (inputValue !== null) {
+      return validateDefaultValue(inputValue, type.ofType, path);
+    }
+    return invalidDefaultValue(
+      `Expected non-nullable type "${inspect(type)}" not to be null.`,
+      path,
+    );
+  }
+
+  if (inputValue === null) {
+    return [];
+  }
+
+  if (isListType(type)) {
+    const itemType = type.ofType;
+    if (isIterableObject(inputValue)) {
+      const errors = [];
+      Array.from(inputValue).forEach((itemValue, index) => {
+        errors.push(
+          ...validateDefaultValue(
+            itemValue,
+            itemType,
+            addPath(path, index, undefined),
+          ),
+        );
+      });
+      return errors;
+    }
+    // Lists accept a non-list value as a list of one.
+    return validateDefaultValue(inputValue, itemType, path);
+  }
+
+  if (isInputObjectType(type)) {
+    if (!isObjectLike(inputValue)) {
+      return invalidDefaultValue(
+        `Expected type "${type.name}" to be an object.`,
+        path,
+      );
+    }
+
+    const errors = [];
+    const fieldDefs = type.getFields();
+
+    for (const field of Object.values(fieldDefs)) {
+      const fieldPath = addPath(path, field.name, type.name);
+      const fieldValue = inputValue[field.name];
+
+      if (fieldValue === undefined) {
+        if (field.defaultValue === undefined && isNonNullType(field.type)) {
+          return invalidDefaultValue(
+            `Field "${field.name}" of required type "${inspect(
+              field.type,
+            )}" was not provided.`,
+            fieldPath,
+          );
+        }
+        continue;
+      }
+
+      errors.push(...validateDefaultValue(fieldValue, field.type, fieldPath));
+    }
+
+    // Ensure every provided field is defined.
+    for (const fieldName of Object.keys(inputValue)) {
+      if (!fieldDefs[fieldName]) {
+        const suggestions = suggestionList(
+          fieldName,
+          Object.keys(type.getFields()),
+        );
+        errors.push(
+          ...invalidDefaultValue(
+            `Field "${fieldName}" is not defined by type "${type.name}".` +
+              didYouMean(suggestions),
+            path,
+          ),
+        );
+      }
+    }
+    return errors;
+  }
+
+  // istanbul ignore else (See: 'https://github.com/graphql/graphql-js/issues/2618')
+  if (isLeafType(type)) {
+    let parseResult;
+    let caughtError;
+
+    // Scalars and Enums determine if a input value is valid via serialize(),
+    // which can throw to indicate failure. If it throws, maintain a reference
+    // to the original error.
+    try {
+      parseResult = type.serialize(inputValue);
```

Comment:
Similar to the other comment: `serialize` is the result coercion. Scalar and Enums check the validity of input values via `parseValue`.